### ISSUE TITLE
resource aci_contract filter option's not saving d_from_port and d_to_port

### ISSUE THREAD

**Issue by scotttyso** [CiscoDevNet/terraform-provider-aci#164]
#138 # Terraform Version
Terraform v0.13.5
+ provider registry.terraform.io/ciscodevnet/aci v0.5.1

### APIC version and APIC Platform
APIC Version is 4.2(5k) - using simulator right now

### Affected Resource(s)
aci_contract

### Terraform Configuration Files

```hcl
resource "aci_contract" "mgmt_ctx" {
	tenant_dn   = aci_tenant.mgmt.id
	description = "Default Mgmt Contract"
	name        = "mgmt_ctx"
	scope       = "tenant"
	filter {
		  description = "Mgmt Traffic"
		  filter_entry {
				entry_description	= "Allow https"
			filter_entry_name   	= "https"
				d_from_port        	= "https"
				d_to_port	        = "https"
				ether_t             = "ipv4"
				prot            	= "tcp"
				stateful			= "yes"  
		  }
		  filter_entry {
				entry_description   = "Allow icmp"
			filter_entry_name   	= "icmp"
				d_from_port        	= "unspecified"
				d_to_port        	= "unspecified"
				ether_t             = "ipv4"
				prot           		= "icmp"
		  }
		  filter_entry {
				entry_description   = "Allow SNMP"
			filter_entry_name   	= "snmp"
				d_from_port        	= "161"
				d_to_port        	= "162"
				ether_t             = "ipv4"
				prot           		= "udp"
		  }
		  filter_entry {
				entry_description   = "Allow ssh"
			filter_entry_name   	= "ssh"
				d_from_port        	= "22"
				d_to_port        	= "22"
				ether_t             = "ipv4"
				prot           		= "tcp" 
				stateful			= "yes" 
		  }
		  filter_name  = "Remote_Mgmt"
	}
}
```

### Debug Output

```hcl
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aci_contract.mgmt_ctx will be updated in-place
  ~ resource "aci_contract" "mgmt_ctx" {
        annotation       = "orchestrator:terraform"
        description      = "Default Mgmt Contract"
        filter_entry_ids = [
            "uni/tn-mgmt/flt-Remote_Mgmt/e-https",
            "uni/tn-mgmt/flt-Remote_Mgmt/e-icmp",
            "uni/tn-mgmt/flt-Remote_Mgmt/e-snmp",
            "uni/tn-mgmt/flt-Remote_Mgmt/e-ssh",
        ]
        filter_ids       = [
            "uni/tn-mgmt/flt-Remote_Mgmt",
        ]
        id               = "uni/tn-mgmt/brc-mgmt_ctx"
        name             = "mgmt_ctx"
        prio             = "unspecified"
        scope            = "tenant"
        target_dscp      = "unspecified"
        tenant_dn        = "uni/tn-mgmt"

      ~ filter {
            description = "Mgmt Traffic"
            filter_name = "Remote_Mgmt"
            id          = "uni/tn-mgmt/flt-Remote_Mgmt"

            filter_entry {
                apply_to_frag     = "no"
                arp_opc           = "unspecified"
                d_from_port       = "443"
                d_to_port         = "443"
                entry_description = "Allow https"
                ether_t           = "ipv4"
                filter_entry_name = "https"
                icmpv4_t          = "unspecified"
                icmpv6_t          = "unspecified"
                id                = "uni/tn-mgmt/flt-Remote_Mgmt/e-https"
                match_dscp        = "unspecified"
                prot              = "tcp"
                s_from_port       = "unspecified"
                s_to_port         = "unspecified"
                stateful          = "yes"
            }
            filter_entry {
                apply_to_frag     = "no"
                arp_opc           = "unspecified"
                d_from_port       = "0"
                d_to_port         = "0"
                entry_description = "Allow icmp"
                ether_t           = "ipv4"
                filter_entry_name = "icmp"
                icmpv4_t          = "unspecified"
                icmpv6_t          = "unspecified"
                id                = "uni/tn-mgmt/flt-Remote_Mgmt/e-icmp"
                match_dscp        = "unspecified"
                prot              = "icmp"
                s_from_port       = "unspecified"
                s_to_port         = "unspecified"
                stateful          = "no"
            }
          ~ filter_entry {
                apply_to_frag     = "no"
                arp_opc           = "unspecified"
              + d_from_port       = "161"
              + d_to_port         = "162"
                entry_description = "Allow SNMP"
                ether_t           = "ipv4"
                filter_entry_name = "snmp"
                icmpv4_t          = "unspecified"
                icmpv6_t          = "unspecified"
                id                = "uni/tn-mgmt/flt-Remote_Mgmt/e-snmp"
                match_dscp        = "unspecified"
                prot              = "udp"
                s_from_port       = "unspecified"
                s_to_port         = "unspecified"
                stateful          = "no"
            }
          ~ filter_entry {
                apply_to_frag     = "no"
                arp_opc           = "unspecified"
              + d_from_port       = "22"
              + d_to_port         = "22"
                entry_description = "Allow ssh"
                ether_t           = "ipv4"
                filter_entry_name = "ssh"
                icmpv4_t          = "unspecified"
                icmpv6_t          = "unspecified"
                id                = "uni/tn-mgmt/flt-Remote_Mgmt/e-ssh"
                match_dscp        = "unspecified"
                prot              = "tcp"
                s_from_port       = "unspecified"
                s_to_port         = "unspecified"
                stateful          = "yes"
            }
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

------------------------------------------------------------------------

This plan was saved to: main.plan

To perform exactly these actions, run the following command to apply:
    terraform apply "main.plan"
```

### Expected Behavior

The problem is that the d_from_port and d_to_port for the second two filter entries are not being saved in the terraform state file... when looking in the state file they are showing as empty ""

So it tried to re-apply the configuration each time.

### Actual Behavior

```hcl
{
      "mode": "managed",
      "type": "aci_contract",
      "name": "mgmt_ctx",
      "provider": "provider[\"registry.terraform.io/ciscodevnet/aci\"]",
      "instances": [
        {
          "schema_version": 1,
          "attributes": {
            "annotation": "orchestrator:terraform",
            "description": "Default Mgmt Contract",
            "filter": [
              {
                "annotation": "",
                "description": "Mgmt Traffic",
                "filter_entry": [
                  {
                    "apply_to_frag": "no",
                    "arp_opc": "unspecified",
                    "d_from_port": "443",
                    "d_to_port": "443",
                    "entry_annotation": "",
                    "entry_description": "Allow https",
                    "entry_name_alias": "",
                    "ether_t": "ipv4",
                    "filter_entry_name": "https",
                    "icmpv4_t": "unspecified",
                    "icmpv6_t": "unspecified",
                    "id": "uni/tn-mgmt/flt-Remote_Mgmt/e-https",
                    "match_dscp": "unspecified",
                    "prot": "tcp",
                    "s_from_port": "unspecified",
                    "s_to_port": "unspecified",
                    "stateful": "yes",
                    "tcp_rules": ""
                  },
                  {
                    "apply_to_frag": "no",
                    "arp_opc": "unspecified",
                    "d_from_port": "0",
                    "d_to_port": "0",
                    "entry_annotation": "",
                    "entry_description": "Allow icmp",
                    "entry_name_alias": "",
                    "ether_t": "ipv4",
                    "filter_entry_name": "icmp",
                    "icmpv4_t": "unspecified",
                    "icmpv6_t": "unspecified",
                    "id": "uni/tn-mgmt/flt-Remote_Mgmt/e-icmp",
                    "match_dscp": "unspecified",
                    "prot": "icmp",
                    "s_from_port": "unspecified",
                    "s_to_port": "unspecified",
                    "stateful": "no",
                    "tcp_rules": ""
                  },
                  {
                    "apply_to_frag": "no",
                    "arp_opc": "unspecified",
                    "d_from_port": "",
                    "d_to_port": "",
                    "entry_annotation": "",
                    "entry_description": "Allow SNMP",
                    "entry_name_alias": "",
                    "ether_t": "ipv4",
                    "filter_entry_name": "snmp",
                    "icmpv4_t": "unspecified",
                    "icmpv6_t": "unspecified",
                    "id": "uni/tn-mgmt/flt-Remote_Mgmt/e-snmp",
                    "match_dscp": "unspecified",
                    "prot": "udp",
                    "s_from_port": "unspecified",
                    "s_to_port": "unspecified",
                    "stateful": "no",
                    "tcp_rules": ""
                  },
                  {
                    "apply_to_frag": "no",
                    "arp_opc": "unspecified",
                    "d_from_port": "",
                    "d_to_port": "",
                    "entry_annotation": "",
                    "entry_description": "Allow ssh",
                    "entry_name_alias": "",
                    "ether_t": "ipv4",
                    "filter_entry_name": "ssh",
                    "icmpv4_t": "unspecified",
                    "icmpv6_t": "unspecified",
                    "id": "uni/tn-mgmt/flt-Remote_Mgmt/e-ssh",
                    "match_dscp": "unspecified",
                    "prot": "tcp",
                    "s_from_port": "unspecified",
                    "s_to_port": "unspecified",
                    "stateful": "yes",
                    "tcp_rules": ""
                  }
                ],
                "filter_name": "Remote_Mgmt",
                "id": "uni/tn-mgmt/flt-Remote_Mgmt",
                "name_alias": ""
              }
            ],
            "filter_entry_ids": [
              "uni/tn-mgmt/flt-Remote_Mgmt/e-https",
              "uni/tn-mgmt/flt-Remote_Mgmt/e-icmp",
              "uni/tn-mgmt/flt-Remote_Mgmt/e-snmp",
              "uni/tn-mgmt/flt-Remote_Mgmt/e-ssh"
            ],
            "filter_ids": [
              "uni/tn-mgmt/flt-Remote_Mgmt"
            ],
            "id": "uni/tn-mgmt/brc-mgmt_ctx",
            "name": "mgmt_ctx",
            "name_alias": "",
            "prio": "unspecified",
            "relation_vz_rs_graph_att": "",
            "scope": "tenant",
            "target_dscp": "unspecified",
            "tenant_dn": "uni/tn-mgmt"
          },
          "private": "eyJzY2hlbWFfdmVyc2lvbiI6IjEifQ==",
          "dependencies": [
            "aci_tenant.mgmt"
          ]
        }
      ]
    }
```

### Steps to Reproduce

1. `terraform apply`

**Comment by conmurphy** [comment]
in this file

https://github.com/CiscoDevNet/terraform-provider-aci/blob/355a01f58f7d7cf8e9a8fdc59b39a2f302abd1e8/aci/resource_aci_vzbrcp.go

starting line 537 in this function `func setFilterEntryAttributesFromContract` there’s a map

```
constantPortMapping := map[string]string{
		"smtp":        "25",
		"dns":         "53",
		"http":        "80",
		"https":       "443",
		"pop3":        "110",
		"rtsp":        "554",
		"ftpData":     "20",
		"ssh":         "22",
		"unspecified": "0",
	}

```

if the `d_from_port` or `d_to_port` keys exist in this map it will return the protocol number. that’s why the https and unspecified filters store the correct state. for your ssh and snmp you’re specifying the protocol numbers which are not matching any keys. therefore it’s returning and storing an empty string

the `d_to__port` and `d_from_port` are using lookups. the `s_from_port` and `s_to_port` are just using the values as you can see on lines 564 to 569

```
	eMap["s_from_port"] = vzEntryMap["sFromPort"]
	eMap["s_to_port"] = vzEntryMap["sToPort"]
	eMap["stateful"] = vzEntryMap["stateful"]
	eMap["tcp_rules"] = vzEntryMap["tcpRules"]
	eMap["d_from_port"] = constantPortMapping[vzEntryMap["dFromPort"]]
	eMap["d_to_port"] = constantPortMapping[vzEntryMap["dToPort"]]
```

not sure on the best fix but maybe a check can be added  and if it doesn’t match then just store the protocol number

**Comment by scotttyso** [comment]
Good enough for me.  I missed it in the docs... thank you for the quick reply.

**Comment by scotttyso** [comment]
You can close this request

**Comment by scotttyso** [comment]
As another thing to change, the only valid protocols are https, smtp, http, ftp-data, Unspecified, dns, pop3, or rtsp... SSH is not a valid option.  That should be removed from the list.  When I just tried this using ssh I got the following:
```
Error: unknown property value ssh, name dFromPort, class vzEntry [(Dn0)] Dn0=, 

  on resources_Tenant_Mgmt.tf line 1, in resource "aci_contract" "mgmt_ctx":
   1: resource "aci_contract" "mgmt_ctx" {
```

**Comment by nkatarmal-crest** [comment]
@scotttyso This issue has been fixed in the v0.5.2 of the provider. Please use that version to run your plan smoothly.
